File: week-3/lectures/merge_sort.py
"""Merge Sort example.

This file belongs to the week 3 lecture examples on graphs, backtracking, and divide-and-conquer algorithms. It uses merge_sort, merge to illustrate the main idea with a small runnable example.
"""

import matplotlib.pyplot as plt
import random

def merge_sort(arr):
    """Sort the list using merge sort.

    Args:
        arr (list): The list being processed by the algorithm.

    Returns:
        list: The list produced by the algorithm.
    """
    if len(arr) > 1:
        mid = len(arr) // 2
        left_arr = arr[:mid]
        right_arr = arr[mid:]
        # recursively sort left and right subarrays
        merge_sort(left_arr)
        merge_sort(right_arr)
        # merge the sorted subarrays
        merge(arr, left_arr, right_arr)

def merge(arr, left_arr, right_arr):
    """Merge two sorted halves into a single sorted result.

    Args:
        arr (list): The list being processed by the algorithm.
        left_arr (list): The sorted left half used during merging.
        right_arr (list): The sorted right half used during merging.

    Returns:
        list: The list produced by the algorithm.
    """
    i = j = k = 0
    while i < len(left_arr) and j < len(right_arr):
        if left_arr[i] < right_arr[j]:
            arr[k] = left_arr[i]
            i += 1
        else:
            arr[k] = right_arr[j]
            j += 1
        k += 1
        # visualize the current state of the array
        plt.bar(range(len(arr)), arr)
        plt.show(block=False)
        plt.pause(0.1)

    while i < len(left_arr):
        arr[k] = left_arr[i]
        i += 1
        k += 1
        # visualize the current state of the array
        plt.bar(range(len(arr)), arr)
        plt.show(block=False)
        plt.pause(0.1)

    while j < len(right_arr):
        arr[k] = right_arr[j]
        j += 1
        k += 1
        # visualize the current state of the array
        plt.bar(range(len(arr)), arr)
        plt.show(block=False)
        plt.pause(0.1)

# generate random data to sort
data = [random.randint(0, 50) for _ in range(15)]
# visualize the initial state
plt.bar(range(len(data)), data)
plt.show(block=False)
plt.pause(1)

# sort the data and visualize each step
merge_sort(data)

# visualize the final sorted state
plt.bar(range(len(data)), data)
plt.show()

File: week-3/lectures/merge_sort.cpp
/**
 * @file merge_sort.cpp
 * @brief Console-based C++ version of the merge-sort visualisation example.
 */

#include <iostream>
#include <vector>

using namespace std;

/**
 * @brief Print the current array state in one row.
 * @param values const vector<int>& List of integers currently being sorted.
 * @return void This helper only prints data and does not return a value.
 */
void print_values(const vector<int>& values) {
    for (int value : values) {
        cout << value << ' ';
    }
    cout << '\n';
}

/**
 * @brief Merge two sorted subarrays into one sorted result.
 * @param values vector<int>& Full list being sorted.
 * @param left int Left boundary of the current merge range.
 * @param mid int Last index of the left half.
 * @param right int Right boundary of the current merge range.
 * @return void This function updates values in place and does not return a value.
 */
void merge(vector<int>& values, int left, int mid, int right) {
    vector<int> merged;
    int left_index = left;
    int right_index = mid + 1;

    while (left_index <= mid && right_index <= right) {
        if (values[left_index] <= values[right_index]) {
            merged.push_back(values[left_index++]);
        } else {
            merged.push_back(values[right_index++]);
        }
    }
    while (left_index <= mid) {
        merged.push_back(values[left_index++]);
    }
    while (right_index <= right) {
        merged.push_back(values[right_index++]);
    }

    for (size_t offset = 0; offset < merged.size(); ++offset) {
        values[left + static_cast<int>(offset)] = merged[offset];
    }
    print_values(values);
}

/**
 * @brief Sort a list with merge sort.
 * @param values vector<int>& List of integers to sort in place.
 * @param left int Left boundary of the current recursive range.
 * @param right int Right boundary of the current recursive range.
 * @return void This function updates values in place and does not return a value.
 */
void merge_sort(vector<int>& values, int left, int right) {
    if (left >= right) {
        return;
    }

    int mid = (left + right) / 2;
    merge_sort(values, left, mid);
    merge_sort(values, mid + 1, right);
    merge(values, left, mid, right);
}

/**
 * @brief Run the merge-sort demonstration.
 * @return int Exit status code returned to the operating system.
 */
int main() {
    vector<int> data = {38, 27, 43, 3, 9, 82, 10, 15, 21, 6, 49, 1, 35, 7, 18};
    print_values(data);
    merge_sort(data, 0, static_cast<int>(data.size()) - 1);
    print_values(data);
    return 0;
}
